### ISSUE TITLE
addC: malformed constraint

### ISSUE THREAD

**Issue by gergoerdi** [ucsd-progsys/liquidhaskell#2369]
Input file:

```
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS -fplugin=LiquidHaskellBoot #-}
module Proxy where

data Proxy (a :: k) = Proxy

tupleToDict :: p (Proxy cs) -> ()
tupleToDict _ = ()

getMaster :: forall p cs. p cs -> ()
getMaster _ = tupleToDict (\x -> x :: Proxy cs)
```

Panic message:

```
input/Proxy.hs:11:1: error:
    Uh oh.
    addC: malformed constraint:
cconsE:
 t = {lq_tmp$x##185 : (GHC.Prim.FUN {lq_tmp$x##177 : 'GHC.Types.Many | $k_##160[VV##159:=lq_tmp$x##177][VV##175:=lq_tmp$x##185]} {lq_tmp$x##179 : ('GHC.Types.BoxedRep {lq_tmp$x##178 : 'GHC.Types.Lifted | $k_##162[VV##161:=lq_tmp$x##178][VV##163:=lq_tmp$x##179][VV##175:=lq_tmp$x##185]}) | $k_##164[VV##163:=lq_tmp$x##179][VV##175:=lq_tmp$x##185]} {lq_tmp$x##181 : ('GHC.Types.BoxedRep {lq_tmp$x##180 : 'GHC.Types.Lifted | $k_##166[VV##165:=lq_tmp$x##180][VV##167:=lq_tmp$x##181][VV##175:=lq_tmp$x##185]}) | $k_##168[VV##167:=lq_tmp$x##181][VV##175:=lq_tmp$x##185]} {lq_tmp$x##184 : (Proxy.Proxy {lq_tmp$x##182 : k | $k_##170[VV##169:=lq_tmp$x##182][VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]} {lq_tmp$x##183 : cs | $k_##172[VV##171:=lq_tmp$x##183][VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]}) | $k_##174[VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]} (Proxy.Proxy {lq_tmp$x##158 : k | $k_##157[VV##156:=lq_tmp$x##158]} {lq_tmp$x##188 : cs | $k_##187[VV##186:=lq_tmp$x##188]})) | $k_##176[VV##175:=lq_tmp$x##185]}
 te = lq_tmp$x##196:{lq_tmp$x##191 : (Proxy.Proxy {lq_tmp$x##189 : k | $k_##138[VV##137:=lq_tmp$x##189][VV##141:=lq_tmp$x##191]} {lq_tmp$x##190 : cs | $k_##140[VV##139:=lq_tmp$x##190][VV##141:=lq_tmp$x##191]}) | $k_##142[VV##141:=lq_tmp$x##191]} -> {lq_tmp$x##194 : (Proxy.Proxy {lq_tmp$x##192 : k | $k_##138[VV##137:=lq_tmp$x##192][VV##141:=lq_tmp$x##194][lq_tmp$x##144:=lq_tmp$x##192][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]} {lq_tmp$x##193 : cs | $k_##140[VV##139:=lq_tmp$x##193][VV##141:=lq_tmp$x##194][lq_tmp$x##145:=lq_tmp$x##193][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]}) | lq_tmp$x##194 == lq_tmp$x##196
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         && $k_##142[VV##141:=lq_tmp$x##194][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]}lq_anf$##7205759403792795640lq_tmp$x##196:{lq_tmp$x##191 : (Proxy.Proxy {lq_tmp$x##189 : k | $k_##138[VV##137:=lq_tmp$x##189][VV##141:=lq_tmp$x##191]} {lq_tmp$x##190 : cs | $k_##140[VV##139:=lq_tmp$x##190][VV##141:=lq_tmp$x##191]}) | $k_##142[VV##141:=lq_tmp$x##191]} -> {lq_tmp$x##194 : (Proxy.Proxy {lq_tmp$x##192 : k | $k_##138[VV##137:=lq_tmp$x##192][VV##141:=lq_tmp$x##194][lq_tmp$x##144:=lq_tmp$x##192][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]} {lq_tmp$x##193 : cs | $k_##140[VV##139:=lq_tmp$x##193][VV##141:=lq_tmp$x##194][lq_tmp$x##145:=lq_tmp$x##193][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]}) | lq_tmp$x##194 == lq_tmp$x##196
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         && $k_##142[VV##141:=lq_tmp$x##194][lq_tmp$x##146:=lq_tmp$x##194][x##avc:=lq_tmp$x##196]}
 <:
{lq_tmp$x##185 : (GHC.Prim.FUN {lq_tmp$x##177 : 'GHC.Types.Many | $k_##160[VV##159:=lq_tmp$x##177][VV##175:=lq_tmp$x##185]} {lq_tmp$x##179 : ('GHC.Types.BoxedRep {lq_tmp$x##178 : 'GHC.Types.Lifted | $k_##162[VV##161:=lq_tmp$x##178][VV##163:=lq_tmp$x##179][VV##175:=lq_tmp$x##185]}) | $k_##164[VV##163:=lq_tmp$x##179][VV##175:=lq_tmp$x##185]} {lq_tmp$x##181 : ('GHC.Types.BoxedRep {lq_tmp$x##180 : 'GHC.Types.Lifted | $k_##166[VV##165:=lq_tmp$x##180][VV##167:=lq_tmp$x##181][VV##175:=lq_tmp$x##185]}) | $k_##168[VV##167:=lq_tmp$x##181][VV##175:=lq_tmp$x##185]} {lq_tmp$x##184 : (Proxy.Proxy {lq_tmp$x##182 : k | $k_##170[VV##169:=lq_tmp$x##182][VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]} {lq_tmp$x##183 : cs | $k_##172[VV##171:=lq_tmp$x##183][VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]}) | $k_##174[VV##173:=lq_tmp$x##184][VV##175:=lq_tmp$x##185]} (Proxy.Proxy {lq_tmp$x##158 : k | $k_##157[VV##156:=lq_tmp$x##158]} {lq_tmp$x##188 : cs | $k_##187[VV##186:=lq_tmp$x##188]})) | $k_##176[VV##175:=lq_tmp$x##185]}
   |
11 | getMaster _ = tupleToDict (\x -> x :: Proxy cs)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by nikivazou** [comment]
It seems that the problem here is that LH does not know that 

`FunTy a b c` is a subtype of `GHC.Prim.FUN a b c`. 

Is this true for GHC? Is [FunTy](https://hackage.haskell.org/package/ghc-9.10.1/docs/src/GHC.Core.TyCo.Rep.html#FunTy) the same as [GHC.Prim.FUN](https://hackage.haskell.org/package/ghc-prim-0.11.0/docs/src/GHC.Prim.html#FUN)?

**Comment by gergoerdi** [comment]
Yes, I believe they are. `FunTy m a b` is "morally" the same as `TyConApp FUN [m, a, b]` but it's used as a special representation inside GHC as an optimization.

**Comment by nikivazou** [comment]
Ok! I believe LH is not aware of this. Will try to see exactly where this information is lost.

**Comment by gergoerdi** [comment]
So I guess this would happen in any situation where the `(->)` type constructor is partially applied? 

Indeed, this also crashes:

```
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS -fplugin=LiquidHaskellBoot #-}
module PartialArrow where

class Cls f where
    meth :: f a a

instance Cls (->) where
    meth = \x -> x
```

**Comment by facundominguez** [comment]
Fixed in #2375
